presto-main/src/main/java/io/prestosql/sql/planner/ExpressionInterpreter.java

```diff
@@ -726,41 +727,93 @@ protected Object visitArithmeticBinary(ArithmeticBinaryExpression node, Object c
         protected Object visitComparisonExpression(ComparisonExpression node, Object context)
         {
             ComparisonExpression.Operator operator = node.getOperator();
+            Expression left = node.getLeft();
+            Expression right = node.getRight();
+
+            if (operator == Operator.IS_DISTINCT_FROM) {
+                return visitIsDistinctFrom(context, left, right);
+            }
+            // Execution engine does not have not equal and greater than operators, so interpret with
+            // equal or less than, but do not flip operator in result, as many optimizers depend on
+            // operators not flipping
+            if (node.getOperator() == Operator.NOT_EQUAL) {
+                Object result = visitComparisonExpression(flipComparison(node), context);
+                if (result == null) {
+                    return null;
+                }
+                if (result instanceof ComparisonExpression) {
+                    return flipComparison((ComparisonExpression) result);
+                }
+                return !(Boolean) result;
+            }
+            if (node.getOperator() == Operator.GREATER_THAN || node.getOperator() == Operator.GREATER_THAN_OR_EQUAL) {
+                Object result = visitComparisonExpression(flipComparison(node), context);
+                if (result instanceof ComparisonExpression) {
+                    return flipComparison((ComparisonExpression) result);
+                }
+                return result;
+            }
 
-            if (operator == ComparisonExpression.Operator.IS_DISTINCT_FROM) {
-                Object left = process(node.getLeft(), context);
-                Object right = process(node.getRight(), context);
+            return visitComparisonExpression(context, operator, left, right);
+        }
 
-                if (left == null && right instanceof Expression) {
-                    return new IsNotNullPredicate((Expression) right);
-                }
+        private Object visitIsDistinctFrom(Object context, Expression leftExpression, Expression rightExpression)
```

Comment:
This is not a visitor method, so let's call it something different such as "processIsDistinctFrom" or "evaluateIsDistinctFrom".